Example command-line project.

MikaelSmith opened the pull request:
An example. It may be overkill.

Commit:
Example command-line project.

Commit:
Add Travis and AppVeyor CI

Commit:
Add minimum acceptance infrastructure for CI

Commit:
CMake and source files derive naming from project

This eliminates all hard-coded names in favor of deriving them from the
project() declaration at the root of the project. Running cmake in a new
project will generate files with the new project's naming scheme.
.gitignore will ignore generated files only for the example project.

pcarlisle commented:
```
% ag facter
vendor/leatherman/scripts/travis_target.sh
23:    # Build cfacter

lib/version.h.in
3:/// Declares macros for the facter library version.
8:/// The facter library major version.
12:/// The facter library minor version.
16:/// The facter library patch version.
21:/// The facter library version as a string (without commit SHA1).
26:/// The facter library version as a string (with commit SHA1).
```

Commit:
Update to latest Leatherman and bump cmake version

Leatherman moved to require cmake 3.2.2; bump cmake version in this
project to also require cmake 3.2.2, and use the project VERSION
property for version number.

MikaelSmith commented:
Latest version should fix all the feedback. Mind giving it a test?

pcarlisle commented on CMakeLists.txt:
```diff
@@ -0,0 +1,64 @@
+set(PROJECT_VERSION_MINOR 0)
+set(PROJECT_VERSION_PATCH 0)
+
+# Comman cmake setup
```
topy

MikaelSmith commented:
The CMake warning will be fixed as soon as https://github.com/puppetlabs/leatherman/pull/33 is merged.

pcarlisle commented:
I still get the pthread error, not sure if that's what you meant by warning since it's a hard error.

MikaelSmith commented:
Nope, I tried something but didn't test it. I'll dig deeper.

MikaelSmith commented:
https://github.com/puppetlabs/leatherman/pull/35 should cover it.

MikaelSmith commented:
Ok, linking errors should be fixed by the work-around commit.

pcarlisle commented:
It all build successfully. I get the following warnings when I run cmake or make

```
CMake Warning (dev) at vendor/leatherman/CMakeLists.txt:2 (project):
  Policy CMP0048 is not set: project() command manages VERSION variables.
  Run "cmake --help-policy CMP0048" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  The following variable(s) would be set to empty:

    PROJECT_VERSION_MAJOR
    PROJECT_VERSION_MINOR
    PROJECT_VERSION_PATCH
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at vendor/leatherman/vendor/boost-nowide/CMakeLists.txt:3 (project):
  Policy CMP0048 is not set: project() command manages VERSION variables.
  Run "cmake --help-policy CMP0048" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  The following variable(s) would be set to empty:

    PROJECT_VERSION_MAJOR
    PROJECT_VERSION_MINOR
    PROJECT_VERSION_PATCH
This warning is for project developers.  Use -Wno-dev to suppress it.
```

Also if I set the project name with a hyphen the resulting binary has converted it to an underscore.

MikaelSmith commented:
I'll look at binary naming. The other is a warning I expect until https://github.com/puppetlabs/leatherman/pull/33 is merged; then I'll bump this project's cmake requirement.

MikaelSmith commented:
Bumped leatherman version to remove the cmake warning, and fixed the Windows compile error.

pcarlisle commented:
It wasn't too hard but I think it'd be useful for future users to comment around the sections of the cmake files that should be deleted once up and running.

pcarlisle commented:
I've been working in earnest from a forked version of this PR for a couple days now and everything seems good to me.

Commit:
Clarify comments and cmake code structure

Adds REMOVE blocks for code that is no longer needed after generating a
project from templates.

MikaelSmith commented:
Cleaned up cmake files, and added REMOVE comments for code that's no longer needed after source has been generated from the template files.

peterhuene commented on exe/driver.cc.in:
```diff
@@ -0,0 +1,89 @@
+        "Usage\n"
+        "=====\n"
+        "\n"
+        "  driver [options]\n"
```
Should this be `@PROJECT_NAME_LOWER@`?

MikaelSmith replied:
Oh, yep. It would be @PROJECT_NAME@, as the executable file can use arbitrary characters. UPPER and LOWER are for macros and namespaces mostly, and corrected to be valid C identifiers.

peterhuene commented on lib/project.cc.in:
```diff
@@ -0,0 +1,16 @@
+
+    string version()
+    {
+        LOG_DEBUG("@PROJECT_NAME@ version is %1%", @PROJECT_NAME_UPPER@_VERSION_WITH_COMMIT);
```
Not sure if logging every time version is requested is useful, but I get the intent.

MikaelSmith replied:
I suppose version is a function that might actually be kept around, but I'd rather do this than create another dummy function to demonstrate logging.

peterhuene replied:
Makes sense :+1: Not hard to remove after template generation.

peterhuene commented:
Other than those small comments, this looks good :+1:

MikaelSmith commented:
Also can't setup coveralls until we make this project public.

MikaelSmith commented:
@peterhuene Updated.

MikaelSmith commented on .travis_target.sh:
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+set -ev
```
I stripped down this file compared to Facter and Leatherman. I'm not sure if I removed anything important (libSegFault?, c++filt?).

peterhuene replied:
Wasn't really used.  I had it there just to make the infrequent segfaults with the initial MRI integration was implemented in libfacter easier to read.
